File: Level-2/73.java
class Solution {
    
    public boolean[] visit;
    public int answer;
    
    public void backTracking(int depth, int k, int[][] dungeons) {
        for(int i = 0; i < dungeons.length; i++) {
            if(!visit[i] && dungeons[i][0] <= k) {
                visit[i] = true;
                backTracking(depth + 1, k - dungeons[i][1], dungeons);
                visit[i] = false;
            }
        }
        answer = Math.max(answer, depth);
    }
    
    public int solution(int k, int[][] dungeons) {
        answer = 0;
        visit = new boolean[dungeons.length];
        
        backTracking(0, k, dungeons);
        
        return answer;
    }
}

File: Level-2/73.js
function solution(k, dungeons) {
    const answer = [];
    const visited = Array(dungeons.length).fill(false);
    
    const dfs = (count, k) => {
        answer.push(count);
        
        for(let i = 0; i < dungeons.length; i++) {
            let current = dungeons[i];
            if(k >= current[0] && !visited[i]) {
                visited[i] = 1;
                dfs(count + 1, k - current[1])
                visited[i] = 0;
            }
        }
    }
    
    dfs(0, k);
    
    return Math.max(...answer);
}
